### DISCUSSION TITLE
1st party support for SwiftUI?

### DISCUSSION THREAD

**Discussion by ursusursus** [airbnb/lottie-ios#2029] [General]
I noticed bunch of articles on the web how to wrap lottie uikit view to swiftui, each with a bit different implementation.

I believe it would be most beneficial if there was 1st party support for this

**Comment by denandreychuk** [comment]
Many developers, myself included, encounter challenges due to the lack of SwiftUI support. We believe this compatibility could significantly enhance the efficacy of the app development process and allow for even more versatile applications

@calda, do you have any plans to add this in the near future?

**Comment by vdotup** [comment]
@calda 
SwiftUI developers create UIKit wrapper for the LottieAnimationView, often with slight variations. It make sense to agree upon the best approach, and integrate it into the official Lottie Package please.

**Comment by calda** [comment]
Thanks all, totally agree this is important. Airbnb has an internal implementation of this that I want to look at moving over to this repo, to serve as the starting point for the open source implementation. That implementation isn't fully-featured though so would still require additional enhancements.

**Reply by denandreychuk** [reply]
Patiently waiting, thanks!😊

**Reply by calda** [reply]
We are starting some work on this here: https://github.com/airbnb/lottie-ios/pull/2086

**Comment by calda** [comment]
Good news, this repo now includes a SwiftUI `LottieView`: https://github.com/airbnb/lottie-ios/pull/2092. We will be extending it with more functionality in follow-ups.

**Reply by denandreychuk** [reply]
Hey @calda , awesome update! Just another community request for future updates - it'd be super cool if progress control with bindings could be a thing. Doing it custom is a bit of a pain. Also, a correct SwiftUI's state invalidation would be nice - sometimes it gets a bit confusing with detecting UIKit views identity, and updating stuff like ColorProviders on the fly gets tricky. Thanks a ton! Keep rocking! 🚀

**Reply by calda** [reply]
Totally. Agree that we will want this to be built-in over time.

**Reply by calda** [reply]
@denandreychuk, here's a PR with new SwiftUI APIs for controlling the animation progress: https://github.com/airbnb/lottie-ios/pull/2108

**Reply by denandreychuk** [reply]
Wow, this is super nice! It appears we are on the verge of using first-party solution in a production environment

**Reply by denandreychuk** [reply]
@calda the last thing that possibly prevents from the transition from custom solutions is the ability to alter animations dynamically.

Consider this use-case: We have multiple animations set up in a sequence, each playing in a loop, one following the other, much like a slideshow.

This presents a couple of challenges:

We need to manage SwiftUI's identity. We need to signal that the animation has changed and that there's an update required in the updateUI of ViewRepresentable.
We need to detect when an animation concludes (completion block) to queue up the next one.
For the first problem, we could potentially address it by applying an explicit id. However, this might lead to the recreation of the view, which could prove costly in terms of resources.

**Reply by denandreychuk** [reply]
I haven't tried it in the Xcode yet, though. Just a quick PR check; maybe it's already solved.

**Reply by calda** [reply]
In this implementation on master you can certainly dynamically update the animation just by changing what `LottieAnimation` you pass in. It doesn't have a robust "play" API that supports completion handlers quite yet, but you can always use the `configure { (view: LottieAnimationView) in ... }` API to call methods directly on the UIView when a state change happens.

**Comment by calda** [comment]
We just released Lottie 4.3.0 which includes official support for SwiftUI: https://github.com/airbnb/lottie-ios/discussions/2189
